File: 0846-Hand of Straights/0846-Hand of Straights.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int W) {
        if (hand.size() % W) {
            return false;
        }
        map<int, int> freq;
        for (int i : hand) {
            ++freq[i];
        }
        int prev = -1, open = 0;
        queue<int> start;
        for (auto it : freq) {
            int curr = it.first;
            int count = it.second;
            if (open > 0 && curr > prev + 1 || open > count) {
                return false;
            }
            start.push(count - open);
            open = count;
            
            if (start.size() == W) {
                open -= start.front();
                start.pop();
            }
            prev = curr;
        }
        return open == 0;
        
    }
};

File: 0801-0900/0846-Hand of Straights/0846-Hand of Straights.py
import collections
class Solution:
    def isNStraightHand(self, hand: List[int], W: int) -> bool:
        freq = collections.Counter(hand)
        prev = -1
        start = collections.deque()
        open = 0
        for curr in sorted(freq):
            count = freq[curr]
            if (open > 0 and curr > prev + 1) or count < open:
                return False
            start.append(count - open)
            open = count

            if len(start) == W:
                open -= start.popleft()

            prev = curr

        return open == 0
